File: python/Count the Number of Complete Components.py
from collections import defaultdict

class Solution:
    def countCompleteComponents(self, n: int, edges: list[list[int]]) -> int:
        graph = defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)

        visited = set()
        complete_components = 0

        def dfs(node, component_nodes):
            stack = [node]
            component_nodes.add(node)
            while stack:
                current = stack.pop()
                for neighbor in graph[current]:
                    if neighbor not in component_nodes:
                        component_nodes.add(neighbor)
                        stack.append(neighbor)

        def is_complete(component_nodes):
            size = len(component_nodes)

            for node in component_nodes:
                if len(graph[node]) != size - 1:
                    return False
            return True

        for node in range(n):
            if node not in visited:
                component_nodes = set()
                dfs(node, component_nodes)
                visited.update(component_nodes)
                if is_complete(component_nodes):
                    complete_components += 1

        return complete_components


# Test cases
if __name__ == "__main__":
    # case 1
    n = 6
    edges = [[0, 1], [0, 2], [1, 2], [3, 4]]
    solution = Solution()
    print("Test Case 1 Output:", solution.countCompleteComponents(n, edges))  # Expected Output: 3

    # case 2
    n = 6
    edges = [[0, 1], [0, 2], [1, 2], [3, 4], [3, 5]]
    print("Test Case 2 Output:", solution.countCompleteComponents(n, edges))  # Expected Output: 1

File: c++/Count the Number of Complete Components.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>

class Solution {
public:
    int countCompleteComponents(int n, std::vector<std::vector<int>>& edges) {
        std::unordered_map<int, std::vector<int>> graph;
        for (const auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        std::unordered_set<int> visited;
        int completeComponents = 0;

        auto dfs = [&](int node, std::unordered_set<int>& componentNodes) {
            std::vector<int> stack = {node};
            componentNodes.insert(node);
            while (!stack.empty()) {
                int current = stack.back();
                stack.pop_back();
                for (int neighbor : graph[current]) {
                    if (componentNodes.find(neighbor) == componentNodes.end()) {
                        componentNodes.insert(neighbor);
                        stack.push_back(neighbor);
                    }
                }
            }
        };

        auto isComplete = [&](const std::unordered_set<int>& componentNodes) {
            int size = componentNodes.size();
            for (int node : componentNodes) {
                if (graph[node].size() != size - 1) {
                    return false;
                }
            }
            return true;
        };

        for (int node = 0; node < n; ++node) {
            if (visited.find(node) == visited.end()) {
                std::unordered_set<int> componentNodes;
                dfs(node, componentNodes);
                visited.insert(componentNodes.begin(), componentNodes.end());
                if (isComplete(componentNodes)) {
                    completeComponents++;
                }
            }
        }

        return completeComponents;
    }
};
// Test case
int main() {
    Solution solution;
    // case 1
    std::vector<std::vector<int>> edges1 = {{0,1}, {0,2}, {1,2}, {3,4}};
    std::cout << "result of test case 1: " << solution.countCompleteComponents(6, edges1) << std::endl; // Expected out put: 3

    // case 2
    std::vector<std::vector<int>> edges2 = {{0,1}, {0,2}, {1,2}, {3,4}, {3,5}};
    std::cout << "result of test case 2: " << solution.countCompleteComponents(6, edges2) << std::endl; // Expected out put: 1

    return 0;
}
